### PR TITLE
[ENG-7487][steps] add support for step input/output params

### PR DESCRIPTION
**Pull request by dsokal**
# Why

This PR adds support for input and output parameters in build config files.

```
build:
  name: Inputs and outputs
  steps:
    - run:
        name: Say HI
        inputs:
          name: Dominik Sokal
        command: echo "Hi, ${ inputs.name }!"
    - run:
        name: Produce output
        id: id123
        outputs: [foo]
        command: |
          echo "Producing output for another step"
          set-output foo bar
    - run:
        name: Produce another output
        id: id456
        outputs:
          - required_param
          - name: optional_param
            required: false
        command: |
          echo "Producing more output"
          set-output required_param abc123
    - run:
        name: Use output from another step
        inputs:
          foo: ${ steps.id123.foo }
          bar: ${ steps.id456.required_param }
          baz: ${ steps.id456.optional_param }
        command: |
          echo "foo = \"${ inputs.foo }\""
          echo "bar = \"${ inputs.bar }\""
          echo "baz = \"${ inputs.baz }\""

```

Input parameters take values that are template string. They can use expressions to access output parameters from another steps. 
Ouput parameters are declared in the step definition and can be set with the `set-output` command available for steps. Output parameters are required by default, but can be set to optional with `required: false`.

The idea behind the `@expo/steps` is that it should be possible to use it in `eas-cli` so parts of the config file validation happen on the client side. This PR is a first step towards implementing the idea. It introduces the build workflow validator that check two things so far:
- if build step IDs are unique, 
- if expressions (interpolation) used in input parameters are valid:
  - they don't reference outputs from future steps
  - they don't reference outputs from non-existent steps
  - they don't reference undefined outputs from existing steps
**The validator does not fail on the first error. Instead, it tries to find all errors so all the user mistakes could be presented with only one pass of the validator.**

# How

Most notable parts of the PR:
- I added a new directory with binaries available to step commands. Currently, it only has the `set-output` command for setting step output. Because of that, I had to change the output directories for TS files from **dist/commonjs** and **dist/esm** to **dist_commonjs** and **dist_esm**, so the path to the **bin** directory is the same for files from **src** and **dist**. 
- I implemented parsing input and output step parameters. Input parameters can take string templates that have expressions that access output parameters from other steps.
- I implemented the first version of the build workflow validator that can show users issues with their build config file on the client side without the need to run the build in the cloud. Currently, it checks for duplicated step ids and invalid usage of interpolation in input parameters.

# Test Plan

- Tests.
- Added new example, **inputs-and-outputs** to the **examples** directory. 
  Screenshot 2023-02-10 at 09 35 00

### CODE REVIEW

File: packages/steps/src/BuildConfigParser.ts [hunk 1]
```diff
@@ -54,4 +78,40 @@ export class BuildConfigParser {
...
+        typeof entry === 'string'
+          ? new BuildStepOutput(this.ctx, { id: entry, stepId, required: true })
+          : new BuildStepOutput(this.ctx, {
+              id: entry.name,
+              stepId,
+              required: entry.required ?? true,
+            });
+      outputs.push(output);
+    }
+    return outputs;
```
Comment (szdziedzic) [inline packages/steps/src/BuildConfigParser.ts:115]:
```suggestion
    return buildStepOutputsConfig.map(entry => {
      typeof entry === 'string'
          ? new BuildStepOutput(this.ctx, { id: entry, stepId, required: true })
          : new BuildStepOutput(this.ctx, {
              id: entry.name,
              stepId,
              required: entry.required ?? true,
            });
          }
```

File: packages/steps/src/__tests__/BuildWorkflowValidator-test.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        }),
+      ],
+    });
+
+    const validator = new BuildWorkflowValidator(workflow);
+    const error = getError<BuildWorkflowError>(() => {
+      validator.validate();
+    });
+    expect(error).toBeInstanceOf(BuildWorkflowError);
+    assert(error instanceof BuildWorkflowError);
```
Comment (szdziedzic) [inline packages/steps/src/__tests__/BuildWorkflowValidator-test.ts:52]:
```suggestion
    expect(error).toBeInstanceOf(BuildWorkflowError);
```

Wouldn't only one of these do the job?

**Review by szdziedzic — APPROVED** [review]
It looks good! 🚀 Good job 🔥 

Thanks for keeping it well-tested and documented!

File: packages/steps/src/BuildWorkflowValidator.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    const notUsedUniqueStepIdsSet = new Set(uniqueStepIds);
+    const duplicatedStepIdsSet = new Set<string>();
+    for (const stepId of stepIds) {
+      if (notUsedUniqueStepIdsSet.has(stepId)) {
+        notUsedUniqueStepIdsSet.delete(stepId);
+      } else {
+        duplicatedStepIdsSet.add(stepId);
+      }
+    }
```
Comment (wkozyra95) [inline packages/steps/src/BuildWorkflowValidator.ts:34]:
```suggestion
    const visitedStepIdsSet = new Set<string>();
    const duplicatedStepIdsSet = new Set<string>();
    for (const stepId of stepIds) {
      if (visitedStepIdsSet.has(stepId)) {
        duplicatedStepIdsSet.add(stepId);
      } else {
        visitedStepIdsSet.add(stepId);
      }
    }
```

No early exit means it's slower, but I think this simplifies the code.
If you prefer to keep your approach, consider renaming notUsedUniqueStepIdsSet

File: packages/steps/src/__tests__/BuildWorkflowValidator-test.ts (hunk 2 above)
Comment (dsokal) [reply packages/steps/src/__tests__/BuildWorkflowValidator-test.ts:52]: No
Screenshot 2023-02-13 at 15 22 52

File: packages/steps/bin/set-output [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+NAME=$1
+VALUE=$2
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/steps/bin/set-output:4]: I added an example and a test.

File: packages/steps/examples/inputs-and-outputs/config.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        command: |
+          echo "Producing output for another step"
+          set-output foo bar
+    - run:
+        name: Produce another output
+        id: id456
+        outputs:
+          - required_param
+          - name: optional_param
+            required: false
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/steps/examples/inputs-and-outputs/config.yml:22]: Testing outputs sounds a bit weird to me, I think only inputs should be validated. If sth failed in a step it should exit with non-zero exit code and I think the author should be responsible for that.

I think your approach can catch some issues with the config a bit earlier, but validating outputs feels wrong.

**Review by wkozyra95 — APPROVED** [review]

**Pull request merged**